cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,446 @@
+              Preconditions.checkArgument(paramAnnotation.containsKey(name),
+                                          String.format("No method parameter found wih a name %s for method %s in " +
+                                                          "class %s whereas it was specified in AuthEnforce " +
+                                                          "annotation.", name, methodName, className));
+            }
+            // Store all the information for the second pass
+            methodAnnotations.put(new Method(methodName, methodDesc),
+                                  new AnnotationDetails(authAnnotation, paramAnnotation));
+          }
+          super.visitEnd();
+        }
+      };
+      return mv;
+    }
+
+    Map<Method, AnnotationDetails> getMethodAnnotations() {
+      return methodAnnotations;
+    }
+  }
+
+  /**
+   * A {@link ClassVisitor} which is used in second pass of {@link AuthEnforceRewriter} to rewrite methods which has
+   * {@link AuthEnforce} annotation on it with the annotation details collected from the first pass. This is only
+   * called for classes which has at least one such method. This class does byte code rewrite to generate call to
+   * {@link AuthEnforceUtil#enforce(AuthorizationEnforcer, EntityId, AuthenticationContext, Set)}. To
+   * see an example of generated byte code please see the documentation for {@link AuthEnforceRewriter}
+   */
+  private final class AuthEnforceAnnotationRewriter extends ClassVisitor {
+
+    private final String className;
+    private final Type classType;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
+
+    AuthEnforceAnnotationRewriter(String className, ClassWriter cw, Map<Method, AnnotationDetails> methodAnnotations) {
+      super(Opcodes.ASM5, cw);
+      this.className = className;
+      this.classType = Type.getObjectType(className.replace(".", "/"));
+      this.methodAnnotations = methodAnnotations;
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, final String methodName, final String methodDesc, String signature,
+                                     String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+      // From the first pass we know the methods which has AuthEnforce annotation and needs to be rewritten
+      if (methodAnnotations.containsKey(new Method(methodName, methodDesc))) {
+        return new AdviceAdapter(Opcodes.ASM5, mv, access, methodName, methodDesc) {
+          @Override
+          protected void onMethodEnter() {
+            // Get the annotation details which was collected by visiting annotations during the first pass
+            AnnotationDetails annotationDetails = methodAnnotations.get(new Method(methodName, methodDesc));
+
+            LOG.debug("AuthEnforce annotation found in class {} on method {}. Authorization enforcement command will " +
```

Comment:
Make this `trace`, otherwise you'll see tons of this message when log level is DEBUG.